client/gc_test.go

```diff
@@ -0,0 +1,325 @@
+	}
+
+	// We should be GC-ing one alloc
+	if gcAlloc := gc.allocRunners.Pop(); gcAlloc == nil {
+		t.Fatalf("err: %v", gcAlloc)
+	}
+
+	if gcAlloc := gc.allocRunners.Pop(); gcAlloc != nil {
+		t.Fatalf("gcAlloc: %v", gcAlloc)
+	}
+}
+
+func TestAllocGarbageCollector_MakeRoomForAllocations_GC_All(t *testing.T) {
+	logger := log.New(os.Stdout, "", 0)
+	statsCollector := &MockStatsCollector{}
+	gc := NewAllocGarbageCollector(logger, statsCollector, 20)
+
+	_, ar1 := testAllocRunnerFromAlloc(mock.Alloc(), false)
+	close(ar1.waitCh)
+	_, ar2 := testAllocRunnerFromAlloc(mock.Alloc(), false)
+	close(ar2.waitCh)
+	if err := gc.MarkForCollection(ar1); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if err := gc.MarkForCollection(ar2); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Make stats collector report 80MB and 95MB free in subsequent calls
+	statsCollector.availableValues = []uint64{80 * MB, 80 * MB, 95 * MB}
+	statsCollector.usedPercents = []float64{0, 0, 0}
+	statsCollector.inodePercents = []float64{0, 0, 0}
+
+	alloc := mock.Alloc()
+	if err := gc.MakeRoomFor([]*structs.Allocation{alloc}); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// We should be GC-ing all the alloc runners
+	if gcAlloc := gc.allocRunners.Pop(); gcAlloc != nil {
+		t.Fatalf("gcAlloc: %v", gcAlloc)
+	}
+}
+
+func TestAllocGarbageCollector_MakeRoomForAllocations_GC_Fallback(t *testing.T) {
+	logger := log.New(os.Stdout, "", 0)
+	statsCollector := &MockStatsCollector{}
+	gc := NewAllocGarbageCollector(logger, statsCollector, 20)
+
+	_, ar1 := testAllocRunnerFromAlloc(mock.Alloc(), false)
+	close(ar1.waitCh)
+	_, ar2 := testAllocRunnerFromAlloc(mock.Alloc(), false)
+	close(ar2.waitCh)
+	if err := gc.MarkForCollection(ar1); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if err := gc.MarkForCollection(ar2); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	alloc := mock.Alloc()
+	if err := gc.MakeRoomFor([]*structs.Allocation{alloc}); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// We should be GC-ing one alloc
+	if gcAlloc := gc.allocRunners.Pop(); gcAlloc == nil {
+		t.Fatalf("err: %v", gcAlloc)
+	}
+
+	if gcAlloc := gc.allocRunners.Pop(); gcAlloc != nil {
+		t.Fatalf("gcAlloc: %v", gcAlloc)
+	}
+}
+
+func TestAllocGarbageCollector_UsageBelowThreshold(t *testing.T) {
+	logger := log.New(os.Stdout, "", 0)
+	statsCollector := &MockStatsCollector{}
+	gc := NewAllocGarbageCollector(logger, statsCollector, 20)
+
+	_, ar1 := testAllocRunnerFromAlloc(mock.Alloc(), false)
+	close(ar1.waitCh)
+	_, ar2 := testAllocRunnerFromAlloc(mock.Alloc(), false)
+	close(ar2.waitCh)
+	if err := gc.MarkForCollection(ar1); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if err := gc.MarkForCollection(ar2); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	statsCollector.availableValues = []uint64{1000}
+	statsCollector.usedPercents = []float64{20}
+	statsCollector.inodePercents = []float64{10}
+
+	if err := gc.keepUsageBelowThreshold(); err != nil {
+		t.Fatalf("err: %v", err)
+	}
```

Comment:
Do these two test anything? Don't they need to check if the allocs have been popped